### PR TITLE
NNX Migration for Mistral models

### PR DESCRIPTION
**Pull request by NuojCheng**
# Description

# Tests

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/mistral.py [hunk 1]
```diff
@@ -59,47 +70,43 @@ def __call__(
...
+    lnx_rms = RMSNorm(
         num_features=inputs.shape[-1],
-        dtype=cfg.dtype,
-        weight_dtype=cfg.weight_dtype,
-        name="pre_self_attention_layer_norm",
+        dtype=self.config.dtype,
+        weight_dtype=self.config.weight_dtype,
         kernel_axes=("norm",),
-        epsilon=cfg.normalization_layer_epsilon,
+        epsilon=self.config.normalization_layer_epsilon,
```
Comment (bvandermoon) [inline MaxText/layers/mistral.py]: Looking good so far, thanks @NuojCheng. You will want to move NNX module initiation like `RMSNorm`, `Attention`, etc. to the `__init__` of `MistralDecoderLayer`

Comment (NuojCheng) [reply MaxText/layers/mistral.py]: Make sense!

File: src/MaxText/layers/mistral.py [hunk 2]
```diff
@@ -21,131 +21,155 @@
...
+from MaxText.common_types import MODEL_MODE_PREFILL
 
 
 # -----------------------------------------
 # The Decoder Layer for Mistral
 # -----------------------------------------
 
 
-class MistralDecoderLayer(nn.Module):
+class MistralDecoderLayerNNX(nnx.Module):
```
Comment (bvandermoon) [inline src/MaxText/layers/mistral.py]: Let's keep the same module name rather than appending everything with NNX

File: src/MaxText/layers/mistral.py [hunk 3]
```diff
@@ -21,131 +21,155 @@
...
+        intermediate_dropout_rate=config.dropout_rate,
+        dtype=config.dtype,
+        weight_dtype=config.weight_dtype,
+        config=config,
+        quant=self.quant,
+        model_mode=model_mode,
+        rngs=self.rngs,
+    )
+
+    self.dropout = nnx.Dropout(rate=config.dropout_rate, broadcast_dims=(-2,), rngs=self.rngs)
```
Comment (bvandermoon) [inline src/MaxText/layers/mistral.py]: Please use the `linears.Dropout` from MaxText instead of `nnx.Dropout`. There are some known issues with the existing `nnx.Dropout` in MaxText

**Review by bvandermoon — COMMENTED** [review]
LGTM just a couple small comments. Are the inference decode test results in the description from after the rebase with #2370? I see some very slight memory diff

Also discussed this offline already, but do you mind running some Maxengine/Jetstream test and loading from a checkpoint?

File: src/MaxText/layers/mistral.py (hunk 3 above)
Comment (NuojCheng) [reply src/MaxText/layers/mistral.py]: Thank you for noticing that. Fixed.

File: src/MaxText/layers/mistral.py (hunk 2 above)
Comment (NuojCheng) [reply src/MaxText/layers/mistral.py]: Fixed.

**Comment by NuojCheng** [comment]
> I see both of accuracy is 0 from JetStream. Did you kill the process or others?
> 
> ```
> Results
> 
> {'accuracy': 0.0, 'gen_num': 5000}
> ```

I did not kill the process. I am not sure what happened here, maybe using the wrong tokenizer? (I used the mistral tokenizer under assets instead of huggingface in llama2)

File: src/MaxText/layers/mistral.py [hunk 4]
```diff
@@ -21,131 +21,155 @@
...
+        model_mode=model_mode,
+        rngs=self.rngs,
+    )
+
+    self.dropout = Dropout(rate=config.dropout_rate, broadcast_dims=(-2,), rngs=self.rngs)
+    
+    if model_mode == MODEL_MODE_PREFILL:
+      self.activation_axis_names = ("activation_batch", "prefill_activation_norm_length", "activation_embed")
+    else:
+      self.activation_axis_names = ("activation_batch", "activation_norm_length", "activation_embed")
```
Comment (bvandermoon) [inline src/MaxText/layers/mistral.py]: Please remove this if it is added based on https://github.com/AI-Hypercomputer/maxtext/pull/2370. Llama2 previously had it in the `__call__` function, so it was just being move to `__init__`. Looks like Mistral doesn't have that

Comment (NuojCheng) [reply src/MaxText/layers/mistral.py]: Thank you. Updated!

**Review by bvandermoon — APPROVED** [review]
LGTM, thanks for the great thorough testing so far. Please do follow up here when you have the results of the golden logits test you mentioned you were running

File: src/MaxText/layers/mistral.py [hunk 5]
```diff
@@ -21,131 +21,151 @@
...
+    self.config = config
+    self.mesh = mesh
+    self.quant = quant
+    self.rngs = rngs
+
+    batch_size, seq_len = max_utils.get_batch_seq_len_for_mode(config, model_mode)
+    dummy_inputs_shape = (batch_size, seq_len, config.emb_dim)
+
+    self.pre_self_attention_layer_norm = RMSNorm(
+        num_features=dummy_inputs_shape[-1],
```
Comment (SurbhiJainUSC) [inline src/MaxText/layers/mistral.py]: Should this be `config.emb_dim`?

Comment (bvandermoon) [reply src/MaxText/layers/mistral.py]: Yes, it should: https://github.com/AI-Hypercomputer/maxtext/pull/2178#discussion_r2283002384

Nice catch @SurbhiJainUSC

Comment (NuojCheng) [reply src/MaxText/layers/mistral.py]: updated. Thank you for the catch!
